### ISSUE TITLE
GCC 12 cannot compile Hyrise

### ISSUE THREAD

**Issue by Bouncner** [hyrise/hyrise#2549]
In order to get GCC running, we need to tackle two issues:
- get the `compact_vector` library updated (PR is open: https://github.com/gmarcais/compact_vector/pull/9)
- use GCC 12.2 (I got `internal compiler errors` when using 12.1) which is unfortunately not shipped with Ubuntu 22.04 as of now
- get Boost Graph running (see below)
- wait for fix of issue with `__builtin_memcpy` (see also here, https://github.com/google/googletest/issues/4108; we might need to wait for 12.3)

**Error log from Boost Graph and GCC 12**
<details>
```
In file included from /usr/include/boost/graph/detail/adjacency_list.hpp:35,
                 from /usr/include/boost/graph/adjacency_list.hpp:255,
                 from /hyrise/src/lib/visualization/abstract_visualizer.hpp:8,
                 from /hyrise/src/lib/visualization/lqp_visualizer.hpp:14,
                 from /hyrise/src/benchmarklib/benchmark_sql_executor.cpp:6:
In member function ‘bool boost::detail::adj_list_edge_iterator<VertexIterator, OutEdgeIterator, Graph>::operator!=(const self&) const [with VertexIterator = boost::range_detail::integer_iterator<long unsigned int>; OutEdgeIterator = boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>; Graph = boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo>]’,
    inlined from ‘void boost::write_graphviz(std::ostream&, const Graph&, VertexPropertiesWriter, EdgePropertiesWriter, GraphPropertiesWriter, VertexID, typename enable_if_c<is_base_and_derived<vertex_list_graph_tag, typename graph_traits<Graph>::traversal_category>::value, graph::detail::no_parameter>::type) [with Graph = adjacency_list<vecS, vecS, directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo>; VertexPropertiesWriter = dynamic_vertex_properties_writer; EdgePropertiesWriter = dynamic_properties_writer; GraphPropertiesWriter = dynamic_graph_properties_writer<adjacency_list<vecS, vecS, directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo> >; VertexID = graph::detail::node_id_property_map<long unsigned int>]’ at /usr/include/boost/graph/graphviz.hpp:284:50:
/usr/include/boost/graph/detail/adj_list_edge_iterator.hpp:112:21: error: ‘*(boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>* const*)((char*)&edge_end + offsetof(boost::detail::edge_iterator, boost::detail::adj_list_edge_iterator<boost::range_detail::integer_iterator<long unsigned int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo, boost::listS> >::edges.boost::optional<std::pair<boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int> > >::<unnamed>.boost::optional_detail::optional_base<std::pair<boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int> > >::m_storage.boost::optional_detail::aligned_storage<std::pair<boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int> > >::dummy_))’ may be used uninitialized [-Werror=maybe-uninitialized]
  111 |                 || (vCurr != vEnd
      |                    ~~~~~~~~~~~~~~
  112 |                     && edges BOOST_GRAPH_MEMBER first
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  113 |                         != x.edges BOOST_GRAPH_MEMBER first);
      |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /hyrise/src/lib/visualization/abstract_visualizer.hpp:9:
/usr/include/boost/graph/graphviz.hpp: In function ‘void boost::write_graphviz(std::ostream&, const Graph&, VertexPropertiesWriter, EdgePropertiesWriter, GraphPropertiesWriter, VertexID, typename enable_if_c<is_base_and_derived<vertex_list_graph_tag, typename graph_traits<Graph>::traversal_category>::value, graph::detail::no_parameter>::type) [with Graph = adjacency_list<vecS, vecS, directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo>; VertexPropertiesWriter = dynamic_vertex_properties_writer; EdgePropertiesWriter = dynamic_properties_writer; GraphPropertiesWriter = dynamic_graph_properties_writer<adjacency_list<vecS, vecS, directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo> >; VertexID = graph::detail::node_id_property_map<long unsigned int>]’:
/usr/include/boost/graph/graphviz.hpp:283:55: note: ‘*(boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>* const*)((char*)&edge_end + offsetof(boost::detail::edge_iterator, boost::detail::adj_list_edge_iterator<boost::range_detail::integer_iterator<long unsigned int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, hyrise::VizVertexInfo, hyrise::VizEdgeInfo, hyrise::VizGraphInfo, boost::listS> >::edges.boost::optional<std::pair<boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int> > >::<unnamed>.boost::optional_detail::optional_base<std::pair<boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int> > >::m_storage.boost::optional_detail::aligned_storage<std::pair<boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int>, boost::detail::out_edge_iter<__gnu_cxx::__normal_iterator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>*, std::vector<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo>, std::allocator<boost::detail::stored_edge_property<long unsigned int, hyrise::VizEdgeInfo> > > >, long unsigned int, boost::detail::edge_desc_impl<boost::directed_tag, long unsigned int>, long int> > >::dummy_))’ was declared here
  283 |     typename graph_traits< Graph >::edge_iterator ei, edge_end;
      |                                                       ^~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-atomic-implicit-seq-cst’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-reserved-id-macro’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-shorten-64-to-32’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-deprecated-dynamic-exception-spec’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-dollar-in-identifier-extension’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
[6/596] Building CXX object src/benchmarklib/CMakeFiles/hyriseBenchmarkLib.dir/benchmark_runner.cpp.o
ninja: build stopped: subcommand failed.

```
</details>

**Comment by Bouncner** [comment]
The `compact_vector` issue will be addressed by #2551.

**Comment by Bouncner** [comment]
To my understanding, several issues are fixed in GCC 13 and Boost might not incorporate fixes for the issues unless 13.1 still fails.
So we might need to wait for GCC 13 (part of non-LTS Ubuntu 23.04) and skip GCC 12 (I don't know if Ubuntu 22.04 will update GCC).

**Comment by Bouncner** [comment]
Fixed with #2611.
